### ISSUE TITLE
Uploaded ClusterConfiguration depends on init phases

### ISSUE THREAD

**Issue by ilia1243** [kubernetes/kubeadm#3029]
## Versions

**kubeadm version** (use `kubeadm version`): v1.29.1

**Environment**:
- **Kubernetes version** (use `kubectl version`): v1.29.1
- **Cloud provider or hardware configuration**: baremetal
- **OS** (e.g. from /etc/os-release): Ubuntu 22.04.1 LTS
- **Kernel** (e.g. `uname -a`):  5.15.0-50-generic
- **Container runtime (CRI)** (e.g. containerd, cri-o): containerd=1.6.12-0ubuntu1~22.04.3
- **Container networking plugin (CNI)** (e.g. Calico, Cilium): Calico
- **Others**:

## What happened?

`kubeadm init` (all phases) and `kubeadm init phase upload-config kubeadm` result to different uploaded `kubeadm-config`. Specifically, in upload-config/kubeadm phase the `kubeadm-config` lacks of authorization-mode parameter.

Most like it is caused by `control-plane` phase and code

```go
cfg.APIServer.ExtraArgs = kubeadmapi.SetArgValues(cfg.APIServer.ExtraArgs, "authorization-mode", getAuthzModes(authzVal), 1)
```

https://github.com/kubernetes/kubernetes/blob/v1.29.2/cmd/kubeadm/app/phases/controlplane/manifests.go#L231

I understand that the issue is cosmetic and does not have real impact.

## What you expected to happen?

ClusterConfiguration is supposed to do not depend on phases.

## How to reproduce it (as minimally and precisely as possible)?

1. `kubeadm init` a single-node cluster.
2. `kubeadm init phase upload-config kubeadm` using the same kubeadm configuration file.

**Comment by neolit123** [comment]
this one might be a "won't fix", because a CC will get mutated along the way during execution of all phases.
if someone has some clever idea how to redesign the code, PRs welcome.

but if this ticket has no takers for a while, i suggest we just close it.

**Comment by ilia1243** [comment]
> CC will get mutated along the way during execution of all phases.

I hope you mean mutating that does not have real impact on subsequent phases and other commands. Otherwise, I feel like I do not get the idea of phrase

> `kubeadm init phase` is consistent with the [kubeadm init workflow](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#init-workflow)

https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init-phase/

It sounds like subsequent execution of all phases one by one should be equivalent to running of the only `kubeadm init` if using the same options.

**Comment by neolit123** [comment]
> > CC will get mutated along the way during execution of all phases.
> 
> I hope you mean mutating that does not have real impact on subsequent phases and other commands. Otherwise, I feel like I do not get the idea of phrase
> 

yes, subsequent phases mutate the CC object for their own needs.

> > `kubeadm init phase` is consistent with the [kubeadm init workflow](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#init-workflow)
> 
> https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init-phase/
> 
> It sounds like subsequent execution of all phases one by one should be equivalent to running of the only `kubeadm init` if using the same options.

we used to have an e2e test for that, but it was removed at some point.
but yes, phases execute one after another and they have the same behavior as the main "init" command, CC is a "state object" that is mutated, and if you only execute one of the phases you might get only one of the states.

**Comment by ilia1243** [comment]
This implies that if execute phases one by one, the uploaded CC might be incomplete by design. Could you confirm one last time that other commands like upgrade are designed to work with incomplete CC the same as with complete (produced by the whole `kubeadm init`)?

**Comment by neolit123** [comment]
> This implies that if execute phases one by one, the uploaded CC might be incomplete by design.

yes, it never was a guarantee to users of the phases.

> Could you confirm one last time that other commands like upgrade are designed to work with incomplete CC the same as with complete (produced by the whole `kubeadm init`)?

upgrade downloads the CC from the CM, generates manifests and applies any custom details from the CC on top of the manifests.
https://github.com/kubernetes/kubernetes/blob/v1.29.2/cmd/kubeadm/app/phases/controlplane/manifests.go#L231 this should be re-applied even if the CM is missing this information. it's part of the defaults.

i'm trying to understand the original problem here:
> kubeadm init (all phases) and kubeadm init phase upload-config kubeadm result to different uploaded kubeadm-config.

can you share the exact diff?

**Comment by ilia1243** [comment]
> this should be re-applied even if the CM is missing this information. it's part of the defaults.

Yes, I have no concerns about the particular `authorization-mode: Node,RBAC` property, it is rather a cosmetic difference. I am concerned about potential impact on other commands from incomplete CC that can be introduced in future versions of kubeadm. I wonder what the approach is, and if it is handled somehow by design.

> can you share the exact diff?

Exactly one property `authorization-mode: Node,RBAC` is removed on `init phase upload-config kubeadm` and no other changes.

> also what flags or --config contents you are passing to init and the phase.

They are the same

<details>

```yaml
apiServer:
  certSANs:
  - extra
  - SANs
  extraArgs:
    admission-control-config-file: /etc/kubernetes/pki/admission.yaml
    audit-log-maxage: '30'
    audit-log-maxbackup: '10'
    audit-log-maxsize: '100'
    audit-log-path: /var/log/kubernetes/audit/audit.log
    audit-policy-file: /etc/kubernetes/audit-policy.yaml
    enable-admission-plugins: NodeRestriction
    profiling: 'false'
  extraVolumes:
  - hostPath: /etc/kubernetes/audit-policy.yaml
    mountPath: /etc/kubernetes/audit-policy.yaml
    name: audit
    pathType: File
    readOnly: true
  - hostPath: /var/log/kubernetes/audit/
    mountPath: /var/log/kubernetes/audit/
    name: audit-log
    pathType: DirectoryOrCreate
    readOnly: false
apiVersion: kubeadm.k8s.io/v1beta3
controlPlaneEndpoint: k8s.example.com:6443
controllerManager:
  extraArgs:
    profiling: 'false'
    terminated-pod-gc-threshold: '1000'
  extraVolumes: []
dns:
  imageRepository: <private.registry>/coredns
imageRepository: <private.registry>
kind: ClusterConfiguration
kubernetesVersion: v1.29.1
networking:
  podSubnet: <podSubnet>
  serviceSubnet: <serviceSubnet>
scheduler:
  extraArgs:
    profiling: 'false'
  extraVolumes: []
---
apiVersion: kubelet.config.k8s.io/v1beta1
cgroupDriver: systemd
enableDebuggingHandlers: true
kind: KubeletConfiguration
podPidsLimit: 4096
protectKernelDefaults: true
readOnlyPort: 0
serializeImagePulls: false
---
apiVersion: kubeproxy.config.k8s.io/v1alpha1
conntrack:
  min: 1000000
kind: KubeProxyConfiguration
---
apiVersion: kubeadm.k8s.io/v1beta3
kind: InitConfiguration
localAPIEndpoint:
  advertiseAddress: <IP>
nodeRegistration:
  criSocket: /var/run/containerd/containerd.sock
  kubeletExtraArgs:
    container-runtime-endpoint: unix:///run/containerd/containerd.sock
  taints: []
patches:
  directory: /etc/kubernetes/patches
```
</details>

**Comment by neolit123** [comment]
looks like the same defaulting bug existing for a while:
https://github.com/kubernetes/kubernetes/commit/f3da0d4ed32fd222057443f285b7c25e71313192#diff-d870edeb820f49abf2f52cd357a8a5396c7cd1a36e024d339146e512186c7297L229-R232

```diff
diff --git a/cmd/kubeadm/app/phases/controlplane/manifests.go b/cmd/kubeadm/app/phases/controlplane/manifests.go
index eb99238fe9b..ae8d2199de5 100644
--- a/cmd/kubeadm/app/phases/controlplane/manifests.go
+++ b/cmd/kubeadm/app/phases/controlplane/manifests.go
@@ -229,7 +229,7 @@ func getAPIServerCommand(cfg *kubeadmapi.ClusterConfiguration, localAPIEndpoint
                cfg.APIServer.ExtraArgs = []kubeadmapi.Arg{}
        }
        authzVal, _ := kubeadmapi.GetArgValue(cfg.APIServer.ExtraArgs, "authorization-mode", -1)
-       cfg.APIServer.ExtraArgs = kubeadmapi.SetArgValues(cfg.APIServer.ExtraArgs, "authorization-mode", getAuthzModes(authzVal), 1)
+       defaultArguments = kubeadmapi.SetArgValues(defaultArguments, "authorization-mode", getAuthzModes(authzVal), 1)
        command = append(command, kubeadmutil.ArgumentsToCommand(defaultArguments, cfg.APIServer.ExtraArgs)...)

```

we can patch it for 1.30 so that the CC is not mutated, but it shouldn't be backported to 1.29 or older - since it's not blocking.
